### PR TITLE
refactor(calibration-storage): Cache robot calibrations and use pydantic models for data validation

### PR DESCRIPTION
**Pull request by Laura-Danielle**
# Overview

Closes RLIQ-118. The objective of this PR is to add in-memory caching for calibration data and pydantic models to support better validation on calibration data in general. Please check out [this confluence doc](https://opentrons.atlassian.net/wiki/spaces/RPDO/pages/3391029314/Calibration+Data+Schemas) for more information on the pydantic schemas and this confluence doc for more information on what the file system looks like in general. There are a lot of loose ends that need to be tied up in separate PRs of which I've documented in <INSERT JIRA EPICS>. 

# Changelog

- Added two modules under _calibration_storage_ to separate OT-2 and OT-3 calibration types
- Cached calibration data in memory to avoid many read/writes to the same file
- Separated instrument calibration data based on OT-2 and OT-3 (pipette(s) will be handled in a follow-up PR which changes the data schema of pipettes to support OT-3 pipette functionality).
- Moved the dataclass objects of the calibration data models (which were only used in the hardware controller) to the hardware controller and converted from the pydantic schemas in the hardware controller once the data was being loaded in.
- Switched to using only pydantic models in the robot-server
- Added more test coverage on the _calibration storage_ module in general
- Modified tests where necessary

# Review requests

Please test this on a robot, let me know if you have any changes you want to the actual data shape as well.

# Test Plan

Please view this confluence document for a more detailed test plan and check list.

- [ ] Run through all OT-2 calibration flows
- [ ] Check that all calibration data endpoints (`calibration/pipette_offset` and `calibation/tip_length`) still work as expected
- [ ] Check that all robot calibrations are still being properly loaded and applied in the hardware controller
- [ ] Check that downgrading the software keeps all calibrations that were calibrated with the new software

# Risk assessment

High. This touches a lot of flakey code and needs a lot of QA eyes. See test plan above.

### CODE REVIEW

File: api/src/opentrons/calibration_storage/ot2/cache.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+from ..types import TiprackHash, PipetteId, TipLengthCalNotFound
+
+from .schemas import v1, defaults
+
+
+TipLengthCalibrations = Dict[PipetteId, Dict[TiprackHash, v1.TipLengthSchema]]
+PipetteCalibrations = Dict[types.MountType, Dict[PipetteId, v1.InstrumentOffsetSchema]]
+
+
+@lru_cache(maxsize=None)
```
Comment (DerekMaggio) [inline api/src/opentrons/calibration_storage/ot2/cache.py:21]: Post from a Slack conversation that I should have asked on here. 

Derek: `Since there are no arguments to your functions will the cache ever be larger than 1 entry?
Because there is only 1 possible argument configuration, and that is no arguments.`

Laura: `We can probably reduce the number, but because we expect the number of times for these functions to be called to be pretty small (i.e. the number of times we are actually updating the cache), it shouldn't matter one way or the other.`

Laura: `Previously, we were doing file lookups for everything so hoping to reduce that number significantly w the cache`

File: api/src/opentrons/calibration_storage/ot2/cache.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+def _tip_length_calibrations() -> TipLengthCalibrations:
+    tip_length_dir = config.get_tip_length_cal_path()
```
Comment (DerekMaggio) [inline api/src/opentrons/calibration_storage/ot2/cache.py:23]: Post from a Slack conversation that I should have asked on here.


Derek: `Could function calls inside the top-level functions cause you to want to calculate a new value, but you will just be looking up the cached value?`

Laura: `In terms of calculating a new value, you'll only want to do that on functions which actually modify the calibrations (which is why I'm clearing the cache in those)`

**Review by DerekMaggio — COMMENTED** [review]
Quite a few of these caching functions are duplicated code. 
Could they be pulled out into shared function file? 

For instance in ot2 and ot3 delete.py you have `_delete_file`, `_remove_json_files_in_directories`, ``clear_tip_length_calibration` ...

At the very least, implementing an interface of shared function that both ot2 and ot3 have to implement could provide some clarity. (I might be getting to Java'y for this)

**Review by mcous — COMMENTED** [review]
Skimmed, but some initial thoughts:

- Should we take this opportunity to make [file access async](https://anyio.readthedocs.io/en/stable/fileio.html)?
- I've found the organizational scheme of `get.py`, `delete.py`, etc. really hard to follow
     - Instead, I strongly feel like they should be organized by calibration data
     - e.g. `tip_length.py`, `deck_attitude.py`, etc.
- I think this PR (source and tests) will be easier to read / understand with much heavier use of `pathlib.Path` rather than `os` and `with open(...) as f:`
- I don't quite understand how / why the `cache.py` is an entirely separate thing
    - I think it would make sense to develop this PR _without_ caching with a good test suite
    - Then, layer on caching as a separate PR, since cache invalidation is hard
    - It seems like the caching strategy we'd want is something that simply puts `lru_cache` on _public_ functions

**Comment by Laura-Danielle** [comment]
Closing in favor of https://github.com/Opentrons/opentrons/pull/11520
